crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -0,0 +1,406 @@
+            self.get_unchecked(index),
+            swap_scratch,
+            self.item_layout.size(),
+        );
+        std::ptr::copy(
+            self.get_unchecked(last),
+            self.get_unchecked(index),
+            self.item_layout.size(),
+        );
+        self.len -= 1;
+        swap_scratch
+    }
+
+    /// # Safety
+    /// index must be in-bounds
+    #[inline]
+    pub unsafe fn swap_remove_unchecked(&mut self, index: usize) {
+        let value = self.swap_remove_and_forget_unchecked(index);
+        (self.drop)(value)
+    }
+
+    /// # Safety
+    /// index must be in-bounds
+    pub unsafe fn swap_remove_type_unchecked<T>(&mut self, index: usize) -> T {
+        let removed = self.swap_remove_and_forget_unchecked(index);
+        std::ptr::read(removed.cast::<T>())
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure this isn't called when len() is 0
+    #[inline]
+    pub unsafe fn pop_unchecked(&mut self) {
+        let ptr = self.get_unchecked(self.len - 1);
+        self.len -= 1;
+        (self.drop)(ptr);
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure the type matches self.item_layout
+    // It is also the caller's responsibility to ensure this isn't called when len() is 0
+    #[inline]
+    pub unsafe fn pop_type_unchecked<T>(&mut self) -> T {
+        let ptr = self.get_unchecked(self.len - 1);
+        self.len -= 1;
+        std::ptr::read(ptr.cast::<T>())
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure that `index` is < self.len()
+    #[inline]
+    pub unsafe fn get_unchecked(&self, index: usize) -> *mut u8 {
+        (*self.data.get())
+            .as_ptr()
+            .add(index * self.item_layout.size())
+    }
+
+    /// Gets a pointer to the start of the vec
+    /// # Safety
+    /// must ensure rust mutability rules are not violated
+    #[inline]
+    pub unsafe fn get_ptr(&self) -> NonNull<u8> {
+        *self.data.get()
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure the type matches self.item_layout and that the type stored is T
+    /// It is also the caller's responsibility to ensure that `index` is < self.len()
+    #[inline]
+    pub unsafe fn get_type_unchecked<T>(&self, index: usize) -> &T {
+        &*self.get_unchecked(index).cast::<T>()
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure the type matches self.item_layout and that the type stored is T
+    /// It is also the caller's responsibility to ensure that `index` is < self.len()
+    #[inline]
+    pub unsafe fn get_type_unchecked_mut<T>(&mut self, index: usize) -> &mut T {
+        &mut *self.get_unchecked(index).cast::<T>()
+    }
+
+    pub fn clear(&mut self) {
+        for i in 0..self.len {
+            unsafe {
+                let ptr = self.get_unchecked(i);
+                (self.drop)(ptr);
+            }
+        }
+
+        self.len = 0;
```

Comment:
```suggestion
        self.len = 0;

        for i in 0..self.len {
            unsafe {
                let ptr = self.get_unchecked(i);
                (self.drop)(ptr);
            }
        }
```

Drop may panic, which could cause `BlobVec` to be dropped. If the length isn't set to zero before any drop, all elements up to the panicking drop will get dropped twice.

Reply:
In this form, wouldn't it drop nothing? You need to store `len` somewhere too.

Reply:
You're right.